### PR TITLE
Raise error on device mismatch in addmm

### LINKED ISSUE #42282: cudaErrorIllegalAddress printing result of torch.nn.Linear(1, 1).cuda()(torch.Tensor([[0.5]]))
## 🐛 Bug

```
>>> import torch
>>> t = torch.Tensor([[0.5]])
>>> model = torch.nn.Linear(1, 1).cuda()
>>> model(t)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/private/home/tbirch/.conda/envs/torch160/lib/python3.7/site-packages/torch/tensor.py", line 153, in __repr__
    return torch._tensor_str._str(self)
  File "/private/home/tbirch/.conda/envs/torch160/lib/python3.7/site-packages/torch/_tensor_str.py", line 371, in _str
    return _str_intern(self)
  File "/private/home/tbirch/.conda/envs/torch160/lib/python3.7/site-packages/torch/_tensor_str.py", line 351, in _str_intern
    tensor_str = _tensor_str(self, indent)
  File "/private/home/tbirch/.conda/envs/torch160/lib/python3.7/site-packages/torch/_tensor_str.py", line 241, in _tensor_str
    formatter = _Formatter(get_summarized_data(self) if summarize else self)
  File "/private/home/tbirch/.conda/envs/torch160/lib/python3.7/site-packages/torch/_tensor_str.py", line 89, in __init__
    nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
RuntimeError: copy_if failed to synchronize: cudaErrorIllegalAddress: an illegal memory access was encountered
```

## Expected behavior

Print out the value of the tensor and not crash

## Environment

Collecting environment information...
PyTorch version: 1.6.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.18.5
[pip3] torch==1.6.0
[pip3] torchvision==0.7.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.1.243             h6bb024c_0
[conda] mkl                       2020.1                      217
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.1.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0da4684_0    conda-forge
[conda] numpy                     1.18.5           py37ha1c710e_0
[conda] numpy-base                1.18.5           py37hde5b4d6_0
[conda] pytorch                   1.6.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] torchvision               0.7.0                py37_cu101    pytorch

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes gh-42282

This adds a device-mismatch check to `addmm` on CPU and CUDA. Although it seems like the dispatcher is always selecting the CUDA version here if any of the inputs are on GPU. So in theory the CPU check is unnecessary, but probably better to err on the side of caution.

### CODE REVIEW

**Comment by ezyang** [comment]
letting @ngimel handle this one

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -199,6 +199,11 @@ static void addmm_impl_cpu_(
   auto m2_strides = m2.strides();
   auto m2_sizes = m2.sizes();
 
+  TORCH_CHECK(self.device() == kCPU && m1.device() == kCPU &&
```
Comment (ngimel) [inline aten/src/ATen/native/LinearAlgebra.cpp:202]: indeed, multidispatch will always send to cuda if one of the tensors is cuda, so this is not needed

File: test/test_cuda.py [hunk 2]
```diff
@@ -3020,5 +3020,29 @@ def test_memory_format_scatter_gather(self):
...
 
 
+    def test_matmul_device_mismatch(self):
+        cpu = torch.rand((10, 10))
+        cuda = cpu.cuda()
+        with self.assertRaisesRegex(RuntimeError, "expected (it|them) to be on GPU"):
+            cpu @ cuda
+        with self.assertRaisesRegex(RuntimeError, "expected (it|them) to be on GPU"):
+            cuda @ cpu
+
```
Comment (ngimel) [inline test/test_cuda.py:3030]: nit: 
for self, m1, m2 in product((cpu,cuda), repeats=3):

**Review by ngimel — COMMENTED** [review]
Looks good, thanks, I have minor comments

**Review by ngimel — APPROVED** [review]

File: test/test_cuda.py [hunk 3]
```diff
@@ -3028,7 +3028,7 @@ def test_matmul_device_mismatch(self):
         with self.assertRaisesRegex(RuntimeError, "expected (it|them) to be on GPU"):
             cuda @ cpu
 
-        for s, m1, m2 in product((cpu, cuda), repeats=3):
+        for s, m1, m2 in product((cpu, cuda), repeat=3):
```
Comment (ngimel) [inline test/test_cuda.py:3031]: lol, sorry :-)

**Pull request closed without merging**